### PR TITLE
CRD Support for flows

### PR DESCRIPTION
**Pull request by sjwoodman**

### CODE REVIEW

File: manager/src/main/java/io/streamzi/openshift/API.java [hunk 1]
```diff
@@ -131,24 +127,21 @@ public String getProcessorFlowDeployment(@PathParam("name") String name) {
     public void postFlow(String flowJson) {
         logger.info(flowJson);
         try {
-            ProcessorFlowReader reader = new ProcessorFlowReader();
-            ProcessorFlow flow = reader.readFromJsonString(flowJson);
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (matzew) [inline manager/src/main/java/io/streamzi/openshift/API.java]: in other methods (e.g. `getGlobalProperties()`) there is also an `ObjectMapper` - let's move it to class level - not method level

File: manager/src/main/resources/flow-crd.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  name: flows.streamzi.io
+spec:
+  group: streamzi.io
+  version: v1
+  scope: Namespaced
+  names:
+    plural: flows
+    singular: flow
+    kind: Flow
+    shortNames:
```
Comment (matzew) [inline manager/src/main/resources/flow-crd.yml]: should we add some 
```
validation:
    openAPIV3Schema:
``` 

for defining the actual structure ?

File: model/src/main/java/io/streamzi/openshift/dataflow/model/ProcessorFlow.java [hunk 3]
```diff
@@ -17,6 +22,31 @@
...
     private Map<String, String> globalSettings = new HashMap<>();
 
+    public ProcessorFlow() {
+    }
+
+    public ProcessorFlow(SerializedFlow flow){
+
+        this.setName(flow.getName());
+        // Add the nodes
+        for(SerializedNode node : flow.getNodes()){
```
Comment (matzew) [inline model/src/main/java/io/streamzi/openshift/dataflow/model/ProcessorFlow.java]: lambdas ?

File: pom.xml [hunk 4]
```diff
@@ -14,7 +14,8 @@
         <failOnMissingWebXml>false</failOnMissingWebXml>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <jackson.version>2.9.6</jackson.version>
-        <openshift.fabric8.version>4.0.3</openshift.fabric8.version>
+        <openshift.fabric8.version>4.0.4</openshift.fabric8.version>
```
Comment (matzew) [inline pom.xml]: is that the fmp ?

File: watcher/src/main/java/io/streamzi/openshift/FlowController.java [hunk 5]
```diff
@@ -27,125 +30,149 @@ public FlowController() {
...
+            for (DeploymentConfig existingDC : existingDCs) {
+                //don't delete the flow itself
+                boolean found = false;
+                for (DeploymentConfig newDC : deploymentConfigs) {
+                    if (existingDC.getMetadata().getName().equals(newDC.getMetadata().getName())) {
+                        found = true;
+                    }
+                }
+
+                if (!found) {
```
Comment (matzew) [inline watcher/src/main/java/io/streamzi/openshift/FlowController.java]: lambda ?

File: watcher/src/main/java/io/streamzi/openshift/FlowController.java [hunk 6]
```diff
@@ -27,125 +30,149 @@ public FlowController() {
...
-                        osClient.deploymentConfigs().inNamespace(osClient.getNamespace()).withName(existingDC.getMetadata().getName()).delete();
+                        logger.info("Deleting ConfigMap: " + existing.getMetadata().getName());
+                        osClient.configMaps().inNamespace(osClient.getNamespace()).withName(existing.getMetadata().getName()).delete();
                     }
                 }
+            }
 
-                //todo: work out why the Strimzi topic CMs aren't being deleted - they get recreated even after the containers have gone away
+            //remove the flow CM if the flow is empty
+            if (flow.getNodes().size() == 0 && flow.getLinks().size() == 0) {
```
Comment (matzew) [inline watcher/src/main/java/io/streamzi/openshift/FlowController.java]: `isEmpty()`

File: watcher/src/main/java/io/streamzi/openshift/FlowController.java [hunk 7]
```diff
@@ -27,125 +30,149 @@ public FlowController() {
...
+                osClient.deploymentConfigs().inNamespace(osClient.getNamespace()).withName(existingDC.getMetadata().getName()).delete();
+            }
 
-            } catch (Exception e) {
-                logger.log(Level.SEVERE, "Error parsing JSON flow data: " + e.getMessage(), e);
+            //todo: work out why the Strimzi topic CMs aren't being deleted - they get recreated even after the containers have gone away
+            List<ConfigMap> existingCMs = osClient.configMaps().inNamespace(osClient.getNamespace()).withLabel("app", flow.getName()).list().getItems();
+            for (ConfigMap existing : existingCMs) {
+                osClient.configMaps().inNamespace(osClient.getNamespace()).withName(existing.getMetadata().getName()).delete();
             }
```
Comment (matzew) [inline watcher/src/main/java/io/streamzi/openshift/FlowController.java]: I think the stuff in the for-loop can be optimised using functional APIs from java8

File: pom.xml (hunk 4 above)
Comment (sjwoodman) [reply pom.xml]: No, it's the fabric8 openshift client :)

File: manager/src/main/resources/flow-crd.yml (hunk 2 above)
Comment (sjwoodman) [reply manager/src/main/resources/flow-crd.yml]: I think we can do this later and re-visit as part of the defining flow structure task.

Comment (matzew) [reply manager/src/main/resources/flow-crd.yml]: +1

**Pull request merged**